### PR TITLE
Index update

### PR DESCRIPTION
**Pull request by jeffreylovitz**
This PR introduces single-property indices for nodes. Persistence is supported, but updates to indices are not. That will be added in a separate PR soon!

### CODE REVIEW

File: demo/imdb/imdb_queries.py [hunk 1]
```diff
@@ -158,6 +158,22 @@
                      ['Judi Dench', '84.000000']]
 )
 
+actors_over_85_index_scan = QueryInfo(
+    query="""MATCH (a:actor)
+             WHERE a.age > 85
+             RETURN a.name, a.age""",
```
Comment (jeffreylovitz) [inline demo/imdb/imdb_queries.py]: MATCH queries do not guarantee order, but the index scan will return values in their stored order (which for this, is increasing age). As `GRAPH.EXPLAIN` won't currently describe index scans, this is valuable right now for making sure the optimization is being used.

File: src/execution_plan/execution_plan.c [hunk 2]
```diff
@@ -93,9 +93,9 @@ void _OpBase_PushBelow(OpBase *a, OpBase *b) {
 
 /* Push b right above a. */
 void _OpBase_PushAbove(OpBase *a, OpBase *b) {
-    /* B shouldn't have its children set. */
```
Comment (jeffreylovitz) [inline src/execution_plan/execution_plan.c]: Changes in these functions all match those introduced in PR #73

File: src/execution_plan/optimizations/utilize_indices.c [hunk 3]
```diff
@@ -0,0 +1,110 @@
+#include "utilize_indices.h"
+#include "../ops/op_index_scan.h"
+
+void _locateScanFilters(NodeByLabelScan *scanOp, Vector *filterOps) {
```
Comment (jeffreylovitz) [inline src/execution_plan/optimizations/utilize_indices.c]: Nearly identical to the `locateTraverse` functions in `translate_filters.c`.

File: demo/imdb/imdb_queries.py (hunk 1 above)
Comment (swilly22) [reply demo/imdb/imdb_queries.py]: We can apply optimisations when the client issued an explain query,
awhile back one of our optimisations required actual data ingestion, but this optimisation was removed.

File: tests/unit/test_skiplist.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+#ifdef __cplusplus
+}
+#endif
+
+char *words[] = {"foo",  "bar",     "zap",    "pomo",
+                 "pera", "arancio", "limone", NULL};
+
+const char *node_label = "default_label";
+char *prop_key = "default_prop_key";
```
Comment (swilly22) [inline tests/unit/test_skiplist.cpp]: Consider turning these globals into class protected variables

File: tests/unit/test_skiplist.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+class SkiplistTest: public ::testing::Test {
+  protected:
+    static void SetUpTestCase() {
+    }
+
+    static void TearDownTestCase() {
+    }
+};
+
+skiplist* build_skiplist(void) {
```
Comment (swilly22) [inline tests/unit/test_skiplist.cpp]: protected function within class

File: tests/unit/test_skiplist.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+// Update key-value pair
+skiplistNode* update_skiplist(skiplist *sl, skiplistVal val, skiplistKey old_key, skiplistKey new_key) {
+  // Ignore the return value from skiplistDelete
+  skiplistDelete(sl, old_key, &val);
+  return skiplistInsert(sl, new_key, val);
+}
+
+void test_skiplist_range(void) {
```
Comment (swilly22) [inline tests/unit/test_skiplist.cpp]: missing TEST_F

File: src/execution_plan/optimizations/utilize_indices.c [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    label = (*scanOp->node)->label;
+    Vector_Clear(filterOps);
+    _locateScanFilters(scanOp, filterOps);
+
+    // No filters.
+    if (Vector_Size(filterOps) == 0) continue;
+
+    /* At this point we have all the filter ops (and thus, filter trees) associated
+     * with the scanned entity. If there are valid indices on any filter and no
+     * equal or higher precedence OR filters, we can switch to an index scan.
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/execution_plan/optimizations/utilize_indices.c]: Right. In that example, the `plan->filterTree` looks like:
```
AND
    OR
        a.age GT 30.000000
        a.age LT 60.000000
    a.gender EQ male
```
And the Vector here has 2 elements:
1:
`a.gender EQ male`
2:
```
OR
    a.age GT 30.000000
    a.age LT 60.000000
```
Though you're right that there's no reason to store non-constant preds, so I'll modify that shortly.

File: src/filter_tree/filter_tree.c [hunk 8]
```diff
@@ -12,10 +12,14 @@
 #include "../query_executor.h"
 #include "../rmutil/vector.h"
 
+// TODO allows for better printing, but may not belong here
+// Lookup map for converting op numbers to strings
+const char* ops[] = {"", "OR", "AND", "ADD", "DASH", "MUL", "DIV", "EQ", "GT", "GE", "LT", "LE"};
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/filter_tree/filter_tree.c]: Ahh, I hadn't thought of that. We can use an enum and a print function, or maybe there's a macro stringizer that could handle this, but I don't think it's a priority - I'll just remove it for right now.

File: src/index/index.c [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+  // Set up a reply of just statistics
+  RedisModule_ReplyWithArray(ctx, 1);
+  RedisModule_ReplyWithArray(ctx, 2);
+
+  RedisModuleKey *key = _index_LookupKey(ctx, graphName, label, prop_str);
+  // Do nothing if this index already exists
+  if (RedisModule_ModuleTypeGetType(key) == IndexRedisModuleType) {
+    RedisModule_CloseKey(key);
+    RedisModule_ReplyWithSimpleString(ctx, "(no changes, no records)");
+    return;
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/index/index.c]: The second reply for all of these is the elapsed time from `module.c:153`, causing CLI outputs like the following:
```
127.0.0.1:6379> GRAPH.QUERY imdb "CREATE INDEX ON :actor(age)"
1) 1) Added 1 index.
   2) "Query internal execution time: 1.319856 milliseconds"
127.0.0.1:6379> GRAPH.QUERY imdb "CREATE INDEX ON :actor(age)"
1) 1) (no changes, no records)
   2) "Query internal execution time: 0.074092 milliseconds"
127.0.0.1:6379> GRAPH.QUERY imdb "DROP INDEX ON :actor(age)"
1) 1) Removed 1 index.
   2) "Query internal execution time: 0.092087 milliseconds"
127.0.0.1:6379> GRAPH.QUERY imdb "DROP INDEX ON :actor(age)"
1) 1) (error) ERR Unable to drop index on :actor(age): no such index.
   2) "Query internal execution time: 0.078346 milliseconds"
```

File: src/index/index.c [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+void Index_Create(RedisModuleCtx *ctx, const char *graphName, Graph *g, const char *label, const char *prop_str) {
+  // Set up a reply of just statistics
+  RedisModule_ReplyWithArray(ctx, 1);
+  RedisModule_ReplyWithArray(ctx, 2);
+
+  RedisModuleKey *key = _index_LookupKey(ctx, graphName, label, prop_str);
+  // Do nothing if this index already exists
+  if (RedisModule_ModuleTypeGetType(key) == IndexRedisModuleType) {
+    RedisModule_CloseKey(key);
+    RedisModule_ReplyWithSimpleString(ctx, "(no changes, no records)");
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/index/index.c]: Yes, that's the verbatim text outputted by the Bolt web connector in that scenario. The various DROP responses match as well. Curiously, the only explicit error for Neo is trying to drop a non-existent index - I'm not sure of the logic, but matched it regardless.

File: tests/unit/test_index.cpp [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+      return g;
+    }
+};
+
+// Define static members of IndexTest class
+const size_t IndexTest::expected_n = 100;
+const char* IndexTest::label = "test_label";
+char* IndexTest::str_key = "string_prop";
+char* IndexTest::num_key = "num_prop";
+int IndexTest::label_id;
```
Comment (jeffreylovitz) [inline tests/unit/test_index.cpp]: I could not find a better way to define static class members (here and in `test_skiplist.cpp`) than this, but I might just be lacking in C++ skills. Tell me if you know a clearer method!

File: src/resultset/resultset.c [hunk 12]
```diff
@@ -304,7 +304,7 @@ ResultSet* NewResultSet(AST_Query* ast, RedisModuleCtx *ctx) {
 }
 
 bool ResultSet_Limited(const ResultSet* set) {
-    return (set->limit != RESULTSET_UNLIMITED);
+    return (set && set->limit != RESULTSET_UNLIMITED);
```
Comment (jeffreylovitz) [inline src/resultset/resultset.c]: This is the only modification required to re-enable optimized EXPLAINs.

File: src/index/index.c [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  RedisModule_CloseKey(key);
+
+  return idx;
+}
+
+void Index_Delete(RedisModuleCtx *ctx, const char *graphName, const char *label, const char *prop) {
+  RedisModule_ReplyWithArray(ctx, 1);
+  RedisModule_ReplyWithArray(ctx, 2);
```
Comment (swilly22) [inline src/index/index.c]: It should be clear for each reply array, where its content is coming from,
either add a comment explaining this, or if possible move reply "declaration" to be in the same scope as the actual reply data.
Please explain reply structure

File: src/index/index.c [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+  }
+}
+
+/* Generate an iterator with no lower or upper bound. */
+IndexIter* IndexIter_Create(Index *idx, SIType type) {
+  skiplist *sl = type == T_STRING ? idx->string_sl : idx->numeric_sl;
+  return skiplistIterateAll(sl);
+}
+
+/* Apply a filter to an iterator, returns 1 if filter can be deleted */
```
Comment (swilly22) [inline src/index/index.c]: Please rephrase: "returns 1 if filter can be deleted"
We won't always be in the context/mindset of SCAN op reduction.

File: src/index/index.h [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+void Index_Delete(RedisModuleCtx *ctx, const char *graphName, const char *label, const char *prop);
+
+/* initializeSkiplists prepares the string and numeric skiplists for a new Index,
+ * pointing them to the appropriate internal comparator routines. It is exposed in
+ * the header so that it can be used by the Index load functions in index_type. */
+void initializeSkiplists(Index *index);
+
+/* buildIndex allocates an Index object and populates it with a label-property pair
+ * by traversing a label matrix with a TuplesIter. */
+Index* buildIndex(Graph *g, const GrB_Matrix label_matrix, const char *label, const char *prop_str);
```
Comment (swilly22) [inline src/index/index.h]: Do we need to expose this function?

File: src/index/index_type.c [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+  uint64_t sl_len = RedisModule_LoadUnsigned(rdb);
+  uint64_t node_num_vals;
+
+  for (int i = 0; i < sl_len; i ++) {
+    // Loop over key-val pairs
+    // TODO ensure this gets freed
+    SIValue *key = malloc(sizeof(SIValue));
+    if (valtype == T_STRING) {
+      *key = SI_StringVal(RedisModule_LoadStringBuffer(rdb, NULL));
+
```
Comment (swilly22) [inline src/index/index_type.c]: Extra space

File: src/index/index_type.c [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+  for (int i = 0; i < sl_len; i ++) {
+    // Loop over key-val pairs
+    // TODO ensure this gets freed
+    SIValue *key = malloc(sizeof(SIValue));
+    if (valtype == T_STRING) {
+      *key = SI_StringVal(RedisModule_LoadStringBuffer(rdb, NULL));
+    
+    } else {
+      *key = SI_DoubleVal(RedisModule_LoadDouble(rdb));
+
```
Comment (swilly22) [inline src/index/index_type.c]: Extra space

File: src/parser/clauses/index.h [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+  CREATE_INDEX
+} AST_IndexOpType;
+
+typedef struct {
+  const char *label;
+  const char *property;
+  AST_IndexOpType operation;
+} AST_IndexNode;
+
+AST_IndexNode* New_AST_IndexNode(const char *label, const char *property, AST_IndexOpType optype);
```
Comment (swilly22) [inline src/parser/clauses/index.h]: Missing a clause validation function.

File: tests/flow/test_imdb.py [hunk 19]
```diff
@@ -152,5 +152,27 @@ def test_find_ten_oldest_actors(self):
         # assert query run time
         self._assert_run_time(actual_result, queries.find_ten_oldest_actors_query)
 
+    def test_index_scan_actors_over_85(self):
+        global redis_graph
+
+        # Execute this command directly, as its response does not contain the result set that
+        # 'redis_graph.query()' expects
+        redis_graph.redis_con.execute_command("GRAPH.QUERY", redis_graph.name, "CREATE INDEX ON :actor(age)")
+        # TODO EXPLAIN queries don't apply optimizations, so this check is disabled for now
```
Comment (swilly22) [inline tests/flow/test_imdb.py]: They do now.

File: tests/unit/test_index.cpp (hunk 11 above)
Comment (swilly22) [reply tests/unit/test_index.cpp]: Why do we need those to be static?

File: tests/unit/test_index.cpp [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+  EXPECT_EQ(str_idx->numeric_sl->length, 0);
+
+  // Build an iterator from a constant filter - this will include all elements
+  SIValue lb = SI_StringVal("");
+  FT_FilterNode *filter = CreateConstFilterNode(label, str_key, GE, lb);
+  IndexIter *iter = IndexIter_Create(str_idx, T_STRING);
+  IndexIter_ApplyBound(iter, &filter->pred);
+  FilterTree_Free(filter);
+
+  GrB_Index *node_id;
```
Comment (swilly22) [inline tests/unit/test_index.cpp]: Let's define a NodeID type as an alias to GrB_Index and use it.

File: tests/unit/test_index.cpp [hunk 21]
```diff
@@ -0,0 +1,266 @@
...
+  Node *cur;
+  SIValue *lb;
+  int ctr = 0;
+  // Find the value of the 10th element in the index
+  while ((node_id = IndexIter_Next(iter)) != NULL) {
+    cur = Graph_GetNode(g, *node_id);
+    lb = GraphEntity_Get_Property((GraphEntity*)cur, num_key);
+    ctr ++;
+    if (ctr == 10) break;
```
Comment (swilly22) [inline tests/unit/test_index.cpp]: if (ctr == 10) {
cur = Graph_GetNode(g, *node_id);
lb = GraphEntity_Get_Property((GraphEntity*)cur, num_key);
break;
}

File: tests/unit/test_skiplist.cpp [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+extern void freeKey(SIValue *key);
+
+#ifdef __cplusplus
+}
+#endif
+
+
+class SkiplistTest: public ::testing::Test {
+  protected:
+    static char *words[];
```
Comment (swilly22) [inline tests/unit/test_skiplist.cpp]: Why did you decide to use static?

File: src/parser/clauses/index.h (hunk 18 above)
Comment (jeffreylovitz) [reply src/parser/clauses/index.h]: What do you think should be checked? The only thing I can think of is that an index clause cannot coexist with any other clause, but I feel like our grammar will enforce this properly.

File: tests/unit/test_index.cpp (hunk 11 above)
Comment (jeffreylovitz) [reply tests/unit/test_index.cpp]: We override GoogleTest's static `SetUpTestCase()` method to instantiate our one graph and two indices just once for all tests. We'll get errors like `invalid use of member ‘IndexTest::str_key’ in static member function` when trying to use non-static member variables.

I feel like there's an obvious approach to this that I'm missing!

File: tests/flow/test_imdb.py [hunk 23]
```diff
@@ -152,5 +152,27 @@ def test_find_ten_oldest_actors(self):
         # assert query run time
         self._assert_run_time(actual_result, queries.find_ten_oldest_actors_query)
 
+    def test_index_scan_actors_over_85(self):
+        global redis_graph
+
+        # Execute this command directly, as its response does not contain the result set that
+        # 'redis_graph.query()' expects
+        redis_graph.redis_con.execute_command("GRAPH.QUERY", redis_graph.name, "CREATE INDEX ON :actor(age)")
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply tests/flow/test_imdb.py]: Agreed. We can support create and drop with a pretty minor change to the `query` logic: if the response from Redis is a one-element array, parse it as query statistics and let the result set not exist. Right now, redisgraph-py anticipates a two-element response, result set and statistics.

We can add explicit functions for create and drop index, but i don't think it's strictly necessary.

File: tests/unit/Makefile [hunk 24]
```diff
@@ -7,7 +7,7 @@ GTEST_DIR = ../../deps/googletest
 CPPFLAGS += -isystem $(GTEST_DIR)/include
 
 # Flags passed to the C++ compiler.
-CXXFLAGS += -g -Wall -Wextra -pthread
+CXXFLAGS += -g -Wall -Wextra -pthread -std=c++11
```
Comment (jeffreylovitz) [inline tests/unit/Makefile]: This (or `gnu++11`) is required for declaring and defining class variables outside of a function body. I don't see any disadvantages to this, but let me know if you have thoughts!

File: src/execution_plan/execution_plan.c [hunk 25]
```diff
@@ -251,9 +251,10 @@ ExecutionPlan* NewExecutionPlan(RedisModuleCtx *ctx, Graph *g,
     QueryGraph *q = NewQueryGraph_WithCapacity(node_count, edge_count);
     execution_plan->graph = q;
     FT_FilterNode *filter_tree = NULL;
-    if(ast->whereNode != NULL)
+    if(ast->whereNode != NULL) {
         filter_tree = BuildFiltersTree(ast->whereNode->filters);
         execution_plan->filter_tree = filter_tree;
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/execution_plan/execution_plan.c]: `execution_plan->filter_tree` is initialized to NULL above here (241).

File: tests/flow/test_imdb.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply tests/flow/test_imdb.py]: Sounds good to me!

File: src/execution_plan/execution_plan.c (hunk 25 above)
[the remark this answers is not in the document]
Comment (swilly22) [reply src/execution_plan/execution_plan.c]: Sorry, missed that.

**Pull request merged**